### PR TITLE
Query Engine multi-granularity support (and MC improvements)

### PR DESCRIPTION
**Pull request by ktsaou**
This PR does the following:

1. Cleans up the query engine to move some conditions inside the query ops of memory modes. This needs to be tested for all memory modes - I have not tested it.
2. Anomaly rates are now calculated ON EVERY data query and the result is part of `RRDR`.
3. Passing the URL option `jw-anomaly-rates` to a data query, Netdata will return BOTH the data and the anomaly rates formatted the same way. So 2 data tables in the result. This is needed on the new charts that will show anomaly rates as a ribbon.
4. Add a new database query grouping function, called `countif`. This function takes an additional parameter that can be given with the URL parameter `group_options`. Read the included README to find out more about it.
5. The `volume` method of metric correlations now uses the `countif` grouping function in case the baseline window is just zero. In this case, it does another query for the highlighted window using `countif` to find the amount of time the highlighted window was not zero. If this does not return any valid results for any reason, it uses the anomaly rate of the highlighted window. If that is not available either, it falls back to the current behavior of returning the average value of the highlighted window. So, now it works with these priorities:
   - the baseline / highlighted ratio, if the average of the baseline window is not zero.
   - if it is zero: the % of time the highlighted window was not zero
   - if this didn't work for any reason: the anomaly rate
   - if that didn't work either: the average of the highlighted window
6. In the case of % of time and anomaly rate, the result is stretched to match the max value of the baseline / highlighted ratio so that the results blend.

All the above are not very well tested currently. @thiagoftsm please help me by testing the non-dbengine memory modes.

The commits have been made with discipline, so they can be reviewed individually. Each one adds one of the above.

## Anomaly Rates in the jasonwrap response:

http://localhost:19999/api/v1/data?context=system.cpu&before=0&after=-7200&points=5&options=jsonwrap,jw-anomaly-rates

```json
{
   "api": 1,
   "id": "system.cpu",
   "name": "system.cpu",
   "view_update_every": 1440,
   "update_every": 1,
   "first_entry": 1654839847,
   "last_entry": 1655385168,
   "before": 1655385120,
   "after": 1655377921,
   "group": "average",
   "options": "jsonwrap,jw-anomaly-rates",
   "dimension_names": ["guest_nice", "guest", "steal", "softirq", "irq", "user", "system", "nice", "iowait"],
   "dimension_ids": ["guest_nice", "guest", "steal", "softirq", "irq", "user", "system", "nice", "iowait"],
   "chart_ids": ["system.cpu", "system.cpu", "system.cpu", "system.cpu", "system.cpu", "system.cpu", "system.cpu", "system.cpu", "system.cpu"],
   "latest_values": [0, 0, 0, 0.1264223, 0.3160556, 2.6548673, 1.1378003, 0.1264223, 0.0632111],
   "view_latest_values": [0, 0, 0, 0.1151908, 0.2700648, 12.962789, 1.5284737, 1.402409, 0.2231769],
   "dimensions": 9,
   "points": 5,
   "format": "json",
   "result": {
 "labels": ["time","guest_nice","guest","steal","softirq","irq","user","system","nice","iowait"],
    "data":
 [
 [ 1655385120,0,0,0,0.1151908,0.2700648,12.962789,1.5284737,1.402409,0.2231769],
 [ 1655383680,0,0,0,0.1320053,0.2670622,6.8018165,1.4328904,0.1651863,0.2355311],
 [ 1655382240,0,0,0,0.2670311,0.3806473,7.5181128,2.682553,0.0836464,0.2449816],
 [ 1655380800,0,0,0,0.185523,0.3578997,6.9362646,1.7946023,0.1442974,0.2362896],
 [ 1655379360,0,0,0,0.2596224,0.5052087,9.733507,2.3730592,0.1810503,0.2293037]
]
},
   "anomaly_rates": {
 "labels": ["time","guest_nice","guest","steal","softirq","irq","user","system","nice","iowait"],
    "data":
 [
 [ 1655385120,0,0,0,0,0,0,0,0,0],
 [ 1655383680,0,0,0,0,0,0.5,0.5,0,0],
 [ 1655382240,0,0,0,0,0,0,1.5,0,0],
 [ 1655380800,0,0,0,0,0,0,1.5,0,0],
 [ 1655379360,0,0,0,0,1,0,0,0,0]
]
},
 "min": 0,
 "max": 12.962789
}
```

The anomaly rate table is percent. 100% there means all points were anomalous.
The granularity of this calculation is 0.5%. This has been done to be optimal on agent resources. We store all these values in `uint8_t` arrays (a single byte per point) with values from 0 to 200 and we only do integer arithmetic in the query engine. Then, when we export it to json, we convert it to a range from 0 to 100, with double precision. So, the granularity is 0.5%.

### CODE REVIEW

**Comment by andrewm4894** [comment]
will test today

**Comment by andrewm4894** [comment]
[some initial testing](https://drive.google.com/file/d/1MlWJGqnxorzH9gdM-Jgg8_YRnuTxQXAC/view?usp=sharing) showing the volume on this branch is much better.

**Comment by andrewm4894** [comment]
[another test](https://drive.google.com/file/d/1nmU5J3HfOv2HA1Mbex8Ptr0ej5zmDR7E/view?usp=sharing) showing that current `volume` is actually maybe broken as it does not really filter anything in the tests ive done:

image

New volume algo from this branch does seem to do much better and at least filters more:

image

ks2 was better than both of them imo but my test here was a big of a sledge hammer so does not really prove anything apart from fact that the `volume` algo on this branch is much better than whats on master.

**Comment by ktsaou** [comment]
@andrewm4894 have another test of this.

MC volume now does 3 queries per dimension:

1. average of the baseline
2. average of the highlighted
3. percentage of time the highlighted window was bigger than the baseline average (when the highlighted average is smaller than the baseline average the check is for smaller, not bigger).

Now the equation is like this:

ratio = (highlight average - baseline average) / baseline average * percentage of time the highlighted window was bigger-smaller than the baseline average

I think this is even better. It penalizes sudden spikes. They are still there, but they blend better in the result.

**Comment by ktsaou** [comment]
In one of your videos you shown that volume was selecting almost all dimensions. This can only happen if most dimensions happen to get the same score (using the equation above). I think it is very unlikely, given that the precision is huge, but still it can happen if you select very small durations (a few seconds) and a lot of dimensions happen to give the same score.

**Comment by ktsaou** [comment]
rebased to master again

**Comment by ktsaou** [comment]
Guys, we only need someone to volunteer testing memory mode `ram`, `map` and `save` trying to do queries outside the boundaries of the database. If this works, this is safe to be merged.

**Comment by stelfrag** [comment]
> Guys, we only need someone to volunteer testing memory mode `ram`, `map` and `save` trying to do queries outside the boundaries of the database. If this works, this is safe to be merged.

It doesn't work with memory mode RAM. I will be able to further investigate later tonight

It will only return a couple of values then `null`

e.g by keep refreshing

```
http://127.0.0.1:19999/api/v1/data?chart=system.cpu&after=-10
```

the response looks like something below

```
{
 "labels": ["time","guest_nice","guest","steal","softirq","irq","user","system","nice","iowait"],
    "data":
 [
 [ 1655476884,0,0.252738,0,0,0,2.358888,2.021904,1.26369,0.42123],
 [ 1655476883,0,0.0834725,0,0.2504174,0,1.919866,2.337229,1.1686144,1.0016694],
 [ 1655476882,null,null,null,null,null,null,null,null,null],
 [ 1655476881,null,null,null,null,null,null,null,null,null],
 [ 1655476880,null,null,null,null,null,null,null,null,null],
 [ 1655476879,null,null,null,null,null,null,null,null,null],
 [ 1655476878,null,null,null,null,null,null,null,null,null],
 [ 1655476877,null,null,null,null,null,null,null,null,null],
 [ 1655476876,null,null,null,null,null,null,null,null,null],
 [ 1655476875,null,null,null,null,null,null,null,null,null]
]
}
```

then on the nest refresh (notice the timestamp 1655476883 is gone)
```
{
 "labels": ["time","guest_nice","guest","steal","softirq","irq","user","system","nice","iowait"],
    "data":
 [
 [ 1655476885,0,0.0840336,0,0.9243697,0,0.5882353,1.1764706,0.9243697,1.764706],
 [ 1655476884,0,0.252738,0,0,0,2.358888,2.021904,1.26369,0.42123],
 [ 1655476883,null,null,null,null,null,null,null,null,null],
 [ 1655476882,null,null,null,null,null,null,null,null,null],
 [ 1655476881,null,null,null,null,null,null,null,null,null],
 [ 1655476880,null,null,null,null,null,null,null,null,null],
 [ 1655476879,null,null,null,null,null,null,null,null,null],
 [ 1655476878,null,null,null,null,null,null,null,null,null],
 [ 1655476877,null,null,null,null,null,null,null,null,null],
 [ 1655476876,null,null,null,null,null,null,null,null,null]
]
}
```

File: database/rrd.h [hunk 1]
```diff
@@ -1211,12 +1211,10 @@ static inline time_t rrdset_slot2time(RRDSET *st, size_t slot) {
         slot = (size_t)st->entries - 1;
     }
 
-    if(slot > rrdset_last_slot(st)) {
-        ret = last_entry_t - (size_t)st->update_every * (rrdset_last_slot(st) - slot + (size_t)st->entries);
-    }
-    else {
-        ret = last_entry_t - (size_t)st->update_every;
```
Comment (ktsaou) [inline database/rrd.h:1218]: Well, this has been wrong, I don't know for how much time.
You cannot find the timestamp of a slot, without using the slot in the equation!

**Comment by ktsaou** [comment]
@stelfrag I did the memory queries support past and future queries. They now track the timestamps as they should and return nulls for both past and future data that do not exist in the db. But I can't find a way to test it. The `allow_past` option does not seem to have any effect on them.

**Comment by ktsaou** [comment]
I think this can be merged. Please review.

**Comment by ktsaou** [comment]
@stelfrag I also moved pack_storage_number() and unpack_storage_number() inside storage engines.

**Comment by ktsaou** [comment]
I also found a bug in the way nulls were filled in storage engine gaps.
The way it was, when the collected value from the storage engine was in the future, the query engine was first using it and then it was filling the gap. But this meant that a value was actually used at the wrong time. It should first fill the gap with nulls and then use the value returned by the storage engine.
I fixed it.

**Comment by ktsaou** [comment]
@stelfrag I also made the default query engine (ex `fixedstep`) support variable step databases and removed the variablestep versions.

The new code, passes the unittests of the variablestep version.

Also, the new code can now generate higher granularity charts than the database (e.g. you have a chart with update every 5, but you want a chart with update every 1 -- yes, it can do that).

This means that all the logic to align points to update every are not needed any more, and once this is done, cloud will be able to aggregate charts from multiple nodes with different granularity each!

**Comment by thiagoftsm** [comment]
Hello @ktsaou ,

I started review your code right now, and I observed that `coverity` did not report errors, but when I compile the branch I am receiving these new warnings:

```sh
web/api/queries/query.c: In function 'rrd2rrdr_do_dimension':
web/api/queries/query.c:466:21: warning: unused variable 'rrd_memory_mode' [-Wunused-variable]
  466 |     RRD_MEMORY_MODE rrd_memory_mode = rd->rrd_memory_mode;
      |                     ^~~~~~~~~~~~~~~
web/api/queries/query.c:460:12: warning: unused variable 'first_time_t' [-Wunused-variable]
  460 |     time_t first_time_t = rrddim_first_entry_t(rd);
      |            ^~~~~~~~~~~~
web/api/queries/query.c:459:12: warning: unused variable 'db_now' [-Wunused-variable]
  459 |     time_t db_now = now;
      |            ^~~~~~
web/api/formatters/json_wrapper.c: In function 'rrdr_json_wrapper_anomaly_rates':
web/api/formatters/json_wrapper.c:359:44: warning: unused parameter 'r' [-Wunused-parameter]
  359 | void rrdr_json_wrapper_anomaly_rates(RRDR *r, BUFFER *wb, uint32_t format, uint32_t options, int string_value) {
      |                                      ~~~~~~^
exporting/process_data.c: In function 'exporting_calculate_value_from_stored_data':
exporting/process_data.c:84:20: warning: unused variable 'n' [-Wunused-variable]
   84 |     storage_number n;
      |                    ^
database/ram/rrddim_mem.c: In function 'rrddim_query_finalize':
database/ram/rrddim_mem.c:85:30: warning: unused variable 'h' [-Wunused-variable]
   85 |     struct mem_query_handle* h = (struct mem_query_handle*)handle->handle;
```

I am going to start tests now with other memory modes and review all commits. :handshake:

**Comment by thiagoftsm** [comment]
A complement for what @stelfrag already reported related to memory mode `RAM`, I can confirm that we are not seeing the charts on top in normal situation:

no_mouse

but, when I moved the mouse on top of a chart, our dashboard is showing the values for that moment:

mouse_on_top

I observed this behavior accessing our dashboard trough `Mozilla Firefox 101.0.1` and `Google Chrome  102.0.5005.115 (Official Build) (64-bit)`

File: database/rrdset.c [hunk 2]
```diff
@@ -1503,6 +1461,7 @@ void rrdset_done(RRDSET *st) {
         // and we have collected metrics for this chart in the past (st->counter != 0)
         // fill the gap (the chart has been just loaded from disk)
         if(unlikely(st->counter) && st->rrd_memory_mode != RRD_MEMORY_MODE_DBENGINE) {
+            // TODO this should be inside the storage engine
```
Comment (thiagoftsm) [inline database/rrdset.c:1464]: Considering this comment, I will give a suggestion that I received in the past in another PR that I made. If this is not moved with this PR, please, create a issue for we never forget this.

**Comment by thiagoftsm** [comment]
Another update related to previous issue that can help us to find the source of the issue, I also observed the same problem for `deduper chart`:

Screenshot_20220620_013440

File: web/api/queries/countif/README.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+The target number and the desired condition can be set using the `group_options` query parameter, as a string, like in these examples:
+
+- `!0`, to match any number except zero.
+- `>=-3` to match any number bigger or equal to -3.
+
+## how to use
+
+This query cannot be used in alarms.
+
+`countif` changes the units of charts. The result of the calculation is always from zero to 1, expressing the percentage of database points that matched the condition.
```
Comment (thiagoftsm) [inline web/api/queries/countif/README.md:30]: I am observing that sometimes I am having `NULL` values in some columns and I always have `NULL` for the most recent value returned when `group_options` is not given:

```sh
$ curl -k https://localhost:19999/api/v1/data?chart=apps.mem&group=countif
{
 "labels": ["time","netdata","apps.plugin","freeipmi.plugin","nfacct.plugin","python.d.plugin","tc-qos-helper","go.d.plugin","ebpf.plugin","httpd","sql","wifi","modem","netmanager","logs","nfs","ssh","dhcp","cron","X","ksmd","system","kernel","other","VMs","time"],
    "data":
 [
 [ 1655689520,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],
 [ 1655689519,0,0,100,0,0,0,0,0,0,0,0,0,0,0,100,0,0,0,0,100,0,100,0,0,100],
 [ 1655689518,0,0,100,0,0,0,0,0,0,0,0,0,0,0,100,0,0,0,0,100,0,100,0,0,100],
 [ 1655689517,0,0,100,0,0,0,0,0,0,0,0,0,0,0,100,0,0,0,0,100,0,100,0,0,100],


 [ 1655689210,0,0,100,0,0,0,0,0,0,0,0,0,0,0,100,0,0,0,0,100,0,100,0,0,0],
 [ 1655689209,0,0,100,0,0,0,0,0,0,0,0,0,0,0,100,0,0,0,0,100,0,100,0,0,100],
 [ 1655689208,0,0,100,0,0,0,0,0,0,0,0,0,0,0,100,0,0,0,0,100,0,100,0,0,null],
 [ 1655689207,0,0,100,0,0,0,0,0,0,0,0,0,0,0,100,0,0,0,0,100,0,100,0,0,null],
 [ 1655689206,0,0,100,0,0,0,0,0,0,0,0,0,0,0,100,0,0,0,0,100,0,100,0,0,null],
```

Another example:

```sh
$ curl -k https://localhost:19999/api/v1/data?chart=filesystem.vfs_deleted_objects&group=countif

{
 "labels": ["time","delete"],
    "data":
 [
 [ 1655690660,null],
 [ 1655690655,0],
```

File: web/api/queries/countif/README.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+- `!0`, to match any number except zero.
+- `>=-3` to match any number bigger or equal to -3.
+
+## how to use
+
+This query cannot be used in alarms.
+
+`countif` changes the units of charts. The result of the calculation is always from zero to 1, expressing the percentage of database points that matched the condition. 
+
+In APIs and badges can be used like this: `&group=countif&group_options=>10` in the URL.
```
Comment (thiagoftsm) [inline web/api/queries/countif/README.md:32]: What I described [here](https://github.com/netdata/netdata/pull/13155/files#r901202168) also happens when I add `group_options`, but I have another concern. 

Firstly I repeat the test given as argument `group_options=>10`:

```sh
$ curl -k https://localhost:19999/api/v1/data?chart=filesystem.vfs_deleted_objects&group=countif&group_options=%3E10
{
 "labels": ["time","delete"],
    "data":
 [
 [ 1655690840,null],
 [ 1655690835,0],
 [ 1655690830,0],
 [ 1655690825,0],
 [ 1655690820,0],
 [ 1655690815,0],
 [ 1655690810,0],
 [ 1655690805,0],
 [ 1655690800,0],
 [ 1655690795,0],
```

After this I modified for a value that does not make any sense `group_options=><=10`:

```sh
$ curl -k https://localhost:19999/api/v1/data?chart=filesystem.vfs_deleted_objects&group=countif&group_options=%3E%3C=10
{
 "labels": ["time","delete"],
    "data":
 [
 [ 1655691000,null],
 [ 1655690995,100],
 [ 1655690990,100],
 [ 1655690985,100],
 [ 1655690980,100],
 [ 1655690975,100],
 [ 1655690970,100],
 [ 1655690965,100],

[ 1655690940,100],
 [ 1655690935,100],
 [ 1655690930,100],
 [ 1655690925,100],
 [ 1655690920,100],
 [ 1655690915,100],
```

For both queries I had the same `200` as answer, but a complete different output:

```sh
2022-06-20 02:07:00: 112872: 21375 '[localhost]:51442' 'DATA' (sent/all = 348/2230 bytes -84%, prep/sent/total = 0.07/0.10/0.17 ms) 200 '/api/v1/data?chart=filesystem.vfs_deleted_objects&group=countif&group_options=>10'
0.07/0.11/0.18 ms) 200 '/api/v1/data?chart=filesystem.vfs_deleted_objects&group=countif&group_options=><=10'
```

Why are we not returning any error when a wrong argument is given?

**Comment by thiagoftsm** [comment]
I am finishing tests today, but I will proceed with them tomorrow, my focus today was related with API, dashboard and streaming, the minor issues that I found I reported.

I also compiled this PR on FreeBSD, and I had netdata running as expected on it, but with analogous issues.

**Comment by ktsaou** [comment]
Fixed all of the mentioned issues.

Pending items:

1. Point duration should be in the past, not in the future
2. db point may not be aligned to query dt
3. update every should not participate in dt calculation of the query engine
4. Once the above are fixed, we need unit tests to make sure they work

**Comment by thiagoftsm** [comment]
Hello @ktsaou ,

I am running the same tests I ran yesterday to confirm the fixes, and I observed that some of the issues were fixed, but there is at least one that is still present. When I query our webserver for apps (`https://localhost:19999/api/v1/data?chart=apps.mem&after=-10&group=countif`) that updates the chart every second everything is fine, but when I query plugins that are not updating every second, like `ebpf.plugin` that updates every 10 seconds, I am still observing some `NULL` values. To identify this issue I ran the following command:

```sh
$ for i in `seq 0 5`; do curl -k -o $i.txt "https://localhost:19999/api/v1/data?chart=filesystem.vfs_deleted_objects&after=-10&group=countif"; sleep 1; done
````

And I got the following results:

| ID | output | Is there `NULL`? |
|-----|----------|------------------------|
|  0  | [0.txt](https://github.com/netdata/netdata/files/8941197/0.txt) | YES |
|  1  | [1.txt](https://github.com/netdata/netdata/files/8941200/1.txt) | YES |
|  2  | [2.txt](https://github.com/netdata/netdata/files/8941205/2.txt) | YES |
|  3  | [3.txt](https://github.com/netdata/netdata/files/8941208/3.txt)  | NO  |
|  4  | [4.txt](https://github.com/netdata/netdata/files/8941210/4.txt)  | NO |
|  5  | [5.txt](https://github.com/netdata/netdata/files/8941215/5.txt)  | YES |

**Comment by thiagoftsm** [comment]
Like the previous comment, when I query for `group=countif&group_options=>10` and `group=countif&group_options=><=10` for charts updated every second the issue is not visible, but for charts that updates every 10 seconds the problems still persists.

I can confirm the majority of the issues were fixed and alignment is still a problem.

Now I am starting another round of tests related to the 4 points described in PR OP.

**Comment by ktsaou** [comment]
For the `null` at the beginning on non-1s-update-every, these following need to finish. Once the following are done, we need to check for invalid output.

> * Point duration should be in the past, not in the future
> * db point may not be aligned to query dt
> * update every should not participate in dt calculation of the query engine
> * Once the above are fixed, we need unit tests to make sure they work

**Comment by ktsaou** [comment]
> The documentation is given correct information for users about the valid values, should we also add an additional information for `unexpected entries`?

Sure. Can you add it?

**Comment by thiagoftsm** [comment]
> > The documentation is given correct information for users about the valid values, should we also add an additional information for `unexpected entries`?
> 
> Sure. Can you add it?

Of course :) , please, give me few hours, because I am doing another round of tests with this PR, when I finish them, I will add a commit to fix it.

**Comment by thiagoftsm** [comment]
Hello @ktsaou ,

I finished tests setting `memory mode` with values `ram`, `debengine`, `save`, and `map`. The behavior of dashoard, API, and heath was the expected. 

I also used our [script](https://github.com/netdata/netdata/tree/master/tests/urls) to do thousands of queries for our charts, due the fact this script still does not have these new features, I also ran queries using the new options for `api/v1/data`. 

I am only waiting for [these](https://github.com/netdata/netdata/pull/13155#issuecomment-1160284449) pending items to run a final round of tests and approve the PR.

**Comment by ktsaou** [comment]
Just pushed an update to fix:

1. Point duration should be in the past, not in the future
2. db point may not be aligned to query dt
3. unit tests work

The following still remains:

4. update every should not participate in dt calculation of the query engine

**Comment by ktsaou** [comment]
I did a commit to move a condition in a better place, to avoid happening without reason on every iteration.
This does not really change anything in the logic.

**Comment by thiagoftsm** [comment]
Hello @ktsaou,

I had another round of tests with your PR. I repeated the previous tests and I focus health today. I imagined initially that I found a problem, but I retest two more times and I observed that the issue I found is also present in master branch. I gave the initial report for @stelfrag .

I am not approving your PR yet, because I am observing sometimes I still have the `NULL` values for the most recent period of time, because data was not stored yet. Are you going to work on this small issue?

Best regards!

**Comment by ktsaou** [comment]
> I am not approving your PR yet, because I am observing sometimes I still have the `NULL` values for the most recent period of time, because data was not stored yet. Are you going to work on this small issue?

Have you observed this on `update_every = 1` charts? It shouldn't happen.

**Review by stelfrag — APPROVED** [review]

**Comment by thiagoftsm** [comment]
> Have you observed this on `update_every = 1` charts? It shouldn't happen on these.

For 1 second it is fine, my concern is related with charts that are updating with higher value.

**Comment by andrewm4894** [comment]
@novykh does/might this have any impact on anomaly rates on every chart stuff? 

The fact that it seems you can get the data and anomaly rates in one call now...i think.

Maybe just something for us to be aware of as opposed to needing to refactor to try take advantage of it (not sure if anywhere we would want to get them both in one call apart from maybe when the AR chart is expanded and we in play mode - but maybe not worth it for now).

**Pull request merged**